### DISCUSSION TITLE
Will "hoist array store" be supported later?

### DISCUSSION THREAD

**Discussion by were** [halide/Halide#1661] [General]
Consider the simple convolution.

````cpp
for (int i = 0; i < 10; ++i) {
    sum[i] = 0;
}
for (int i = 0; i < 10; ++i) {
    for (int j = 0; j < 3; ++i) {
        sum[i] = sum[i] + a[i+j] * k[j];
    }
}
for (int i = 0; i < 10; ++i) {
    conv[i] = sum[i]
}
````

In the second stage of computation, the `Store` of array `sum` can be hoisted.
````cpp
for (int i = 0; i < 10; ++i) {
    int _sum = sum[i];
    for (int j = 0; j < 3; ++i) {
        _sum = _sum + a[i+j] * k[j];
    }
    sum[i] = _sum;
}
````

**Comment by abadams** [comment]
This is a different loop nest structure (the three outer loops have been fused), so you get at it using scheduling. We never change the loop structure specified by the schedule, so this optimization is not something Halide should do automatically. You want:

sum.compute_at(conv, i)

If there's no wrapper function to schedule it within, and it's called directly, you can use Func::in to make a wrapper:

sum.compute_at(sum.in(), i);

**Comment by were** [comment]
It a new scheduling function, right?
I haven't updated my Halide for a long time.
Sorry.

**Comment by abadams** [comment]
Func::in is new. compute_at has existed a long time.
